openfl/federated/data/sources/torch/s3_folder.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020-2025 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from urllib.parse import urlparse
+
+import boto3
+
+from openfl.federated.data.sources.torch.folder_dataset import FolderDataset, LabelMapper
+
+
+class S3Folder(FolderDataset):
+    def __init__(
+        self,
+        uri,
+        label_mapper: LabelMapper,
+        endpoint=None,
+        access_key_env_name=None,
+        secret_key_env_name=None,
+        transform=None,
+    ):
+        """
+        Args:
+            uri (str): URI to the S3 object.
+            label_mapper (LabelMapper): LabelMapper object to map class names to indices.
+            endpoint (str, optional): S3 endpoint URL.
+            access_key_env_name (str, optional): Environment variable name for S3 access key.
+            secret_key_env_name (str, optional): Environment variable name for S3 secret key.
+            transform (callable, optional): Transformations to apply to images.
+        """
+        self.uri = uri
+        self.endpoint = endpoint
+        self.access_key_env_name = access_key_env_name
+        self.secret_key_env_name = secret_key_env_name
+        super().__init__(label_mapper, transform=transform)
+
+    def _load_samples(self):
+        """Loads all file paths and their inferred labels into self.samples."""
+        parsed = urlparse(self.uri)
+        bucket_name = parsed.netloc
+        prefix = parsed.path.lstrip("/")
+        access_key = os.environ.get(self.access_key_env_name) if self.access_key_env_name else None
+        secret_key = os.environ.get(self.secret_key_env_name) if self.secret_key_env_name else None
+
+        s3_client = boto3.client(
+            "s3",
+            endpoint_url=self.endpoint,
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+        paginator = s3_client.get_paginator("list_objects_v2")
+        samples = []
```

Comment:
There seems to be some duplication with the logic in `S3DataSource`. Couldn't the `S3Folder` reference a `S3DataSource` instead of directly interacting with S3?

Reply:
Good point, I added a small refactor in a separate commit